Fix app-layer protocol detect patterns for SSLv2

giannitedesco opened the pull request:
- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/issues/4657) ticket:

The SSLv2 record format starts with 2 byte length, then it has a 1 byte
message-type field set to |01| for the client handshake, then a 2 byte
protocol version.

The SSLv3 record format (in use with modern TLS) starts with the
message-type of |16| and then the protocol, followed by the length.

This means that several of the patterns in the SSL app-layer detection
code are incorrect because they're looking in the wrong place in the
packet.

suricata-verify-pr: 555
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

victorjulien commented:
Are you able to provide a Suricata-Verify test?

victorjulien commented:
I see this targets `master-6.0.x`. Please target master instead, and we'll get a backport in afterwards.

giannitedesco commented:
Yes, I think I can provide a test case. I will look in to it on Monday. Rebased it on to master in the meantime.

victorjulien commented:
@giannitedesco did you find a test case to share?

giannitedesco commented:
I have one, I'm just on vacation right now so it'll be another week or two
until i can get it all together!

On Wed, Sep 29, 2021, 21:50 Victor Julien ***@***.***> wrote:

> @giannitedesco <https://github.com/giannitedesco> did you find a test
> case to share?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OISF/suricata/pull/6326#issuecomment-930144271>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABWBL3ZFSKY47CXR4O5QP3UEMDQFANCNFSM5DHU4DZQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

giannitedesco commented:
@victorjulien ready with the tests now, I added 4 tests where 3/4 work before this patch and 4/4 work after.

Maybe you can help me out though? What's the easiest way to enable tracing of the SSL/TLS decoder so I can get a clue of why it doesn't parse the handshakes? I'd like to dig in to it to see if my pcaps are bad, or if there are other bugs in the decoder or my suricata.yaml

victorjulien commented:
> Okay, so I've discovered the issue, there's two:
> 
> 1. for SSLv2 the server handshake, and therefore the certificate, are not parsed. That prevents us from generating the `TLS_HANDSHAKE_DONE` event which would trigger the eve log.
> 2. for SSLv2 client handshake upgrading to TLS, my pcap doesn't contain what I thought it did, it should contain the server TLS handshake.
> 
> I will fix the latter issue with the test, and not sure what to do about the former issue. Should SSLv2 events show up in the log? Looks like it needs to be implemented, but it's quite hard to generate real SSLv2 traffic these days. Maybe it's not worth it?

I can't find it now of course, but I recently read an article show that there is a surprisingly long tail of old ssl and tls versions in use still. So SSLv2 support may still be worth it. That said, if we're missing bigger features I don't think that should block us from already fixing things. Wrt the logging can you point this out in a test case?

giannitedesco commented:
verify PR is OISF/suricata-verify/pull/555

catenacyber commented:
Could you fix the formatting ? cf `./scripts/clang-format.sh`

catenacyber reviewed (changes requested):
Could you fix the formatting ? cf ./scripts/clang-format.sh

giannitedesco commented:
> Could you fix the formatting ? cf ./scripts/clang-format.sh

```
$ ./scripts/clang-format.sh check-branch --diff
clang-format did not modify any files

First commit on branch: 38560d40fe9c56fd869e5a9bf1c9f80e710e19bb
no modified files to format
```

All seems to be fine?

giannitedesco commented:
Seems clang-check isn't actually failing on github either, Looks like it's just random build failures...?

catenacyber commented:
> Seems clang-check isn't actually failing on github either, Looks like it's just random build failures...?

Does not look like the builds failures I know...

Locally I get
```
./scripts/clang-format.sh check-branch
First commit on branch: 38560d40fe9c56fd869e5a9bf1c9f80e710e19bb
ERROR: Branch requires formatting
ERROR: View required changes with: clang-format.sh check-branch --diff
ERROR: Use clang-format.sh rewrite-branch or clang-format.sh branch to fix formatting
```

The diff is like 
```
diff --git a/src/app-layer-ssl.c b/src/app-layer-ssl.c
index 1b27fb738..bfa2e6377 100644
--- a/src/app-layer-ssl.c
+++ b/src/app-layer-ssl.c
@@ -2770,9 +2770,8 @@ static int SSLRegisterPatternsForProtocolDetection(void)
     }
 
     /** SSLv3 */
-    if (AppLayerProtoDetectPMRegisterPatternCS(IPPROTO_TCP, ALPROTO_TLS,
-                                               "|01 03 00|", 5, 2, STREAM_TOSERVER) < 0)
-    {
+    if (AppLayerProtoDetectPMRegisterPatternCS(
+                IPPROTO_TCP, ALPROTO_TLS, "|01 03 00|", 5, 2, STREAM_TOSERVER) < 0) {
         return -1;
     }
```

Maybe try to rebase on latest master...
Could you also check the Suricata-verify failures ?
```
===> sslv2-handshake: Sub test #2: FAIL : expected 1 matches; got 0 for filter {'count': 1, 'match': {'event_type': 'tls', 'tls.serial': '01:A3:B2:E2:6F:04:5B:78:B9:C6:95:C0:1D:87:96:3B:01:61:F4:EA', 'tls.issuerdn': 'C=XX, L=Default City, O=Default Company Ltd'}}
21
===> sslv2-offport-handshake: Sub test #2: FAIL : expected 1 matches; got 0 for filter {'count': 1, 'match': {'event_type': 'tls', 'tls.serial': '01:A3:B2:E2:6F:04:5B:78:B9:C6:95:C0:1D:87:96:3B:01:61:F4:EA', 'tls.issuerdn': 'C=XX, L=Default City, O=Default Company Ltd'}}
```

giannitedesco commented:
Ah, the clang script modified the code, I updated it now. Rebased to master.

The suricata-verify failures are correct. There is no support for SSLv2... It needs to be added if those tests are ever to succeed.

giannitedesco commented:
Welp, still getting weird build failures. No idea...

catenacyber commented:
Looks like you did not pull the latest `master` branch

Maybe this is an issue with GitHub actions cache.
You can try to close this PR and open a new one...
